### PR TITLE
Fix metabase-lib/metadata/Metadata type

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/30763

Further refactors `getMetadata` selector to be typecast free and type safe. Also fixes types in `Metadata` class and removes typescript ignore comments. For `getMetadata`, I've moved hydration code to separate functions, and fixed cases where it relied on hydrated properties being overwritten by using `getPlainObject` explicitly. 

Next steps:
- Move table and filter filtering logic to the selector body
- Move selector body to `metabase-lib`, like `Metadata.create({ databases, ... })` or something like that

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30808)

### CODE REVIEW

File: frontend/src/metabase/selectors/metadata.ts [hunk 1]
```diff
@@ -103,103 +108,61 @@ export const getMetadata: (
...
-      "remapping",
-      field => new Map(getRemappings(field)),
+    metadata.metrics = Object.fromEntries(
+      Object.values(metrics).map(m => [m.id, createMetric(m, metadata)]),
+    );
+    metadata.questions = Object.fromEntries(
+      Object.values(questions).map(c => [c.id, createQuestion(c, metadata)]),
     );
 
+    Object.values(metadata.databases).forEach(database => {
```
Comment (ranquild) [inline frontend/src/metabase/selectors/metadata.ts:133]: The hydration order hasn't changed

File: frontend/src/metabase/selectors/metadata.ts [hunk 2]
```diff
@@ -64,17 +67,19 @@ const getNormalizedFields = createSelector(
     getIncludeSensitiveFields,
   ],
   (fields, tables, includeHiddenTables, includeSensitiveFields) =>
-    filterValues(fields, field => {
-      const table = tables[field.table_id];
+    Object.fromEntries(
+      Object.entries(fields).filter(([, field]) => {
```
Comment (uladzimirdev) [inline frontend/src/metabase/selectors/metadata.ts:71]: filterValues seems to be easier 😅

**Review by uladzimirdev — APPROVED** [review]

File: frontend/src/metabase/selectors/metadata.ts [hunk 3]
```diff
@@ -249,71 +219,142 @@ function createField(field: NormalizedField, metadata: Metadata) {
...
+    : schema.database && schema.database.getTables().length > 0
+    ? // if the schema has a database with tables, use those
+      schema.database
+        .getTables()
+        .filter(table => table.schema_name === schema.name)
+    : // otherwise use any loaded tables that match the schema id
+      Object.values(metadata.tables).filter(
+        table => table.schema && table.schema.id === schema.id,
+      );
+}
```
Comment (kulyk) [inline frontend/src/metabase/selectors/metadata.ts:292]: This is pretty hard to read, can we avoid nested ternaries?

File: frontend/src/metabase/selectors/metadata.ts [hunk 4]
```diff
@@ -220,25 +183,32 @@ export const getMetadataWithHiddenTables = (
 
 // Utils
 
-function createDatabase(db: NormalizedDatabase, metadata: Metadata) {
-  const instance = new Database(db);
+function isNotNull<T>(value: T | null | undefined): value is T {
+  return value != null;
+}
```
Comment (kulyk) [inline frontend/src/metabase/selectors/metadata.ts:188]: Should we use `isNotNull` from `metabase/core/utils/types`?

**Review by kulyk — APPROVED** [review]
Overall looks good, left a few minor-ish comments

File: frontend/src/metabase/selectors/metadata.ts [hunk 5]
```diff
@@ -277,18 +277,17 @@ function hydrateSchemaDatabase(
...
-    : schema.database && schema.database.getTables().length > 0
-    ? // if the schema has a database with tables, use those
-      schema.database
-        .getTables()
-        .filter(table => table.schema_name === schema.name)
-    : // otherwise use any loaded tables that match the schema id
-      Object.values(metadata.tables).filter(
-        table => table.schema && table.schema.id === schema.id,
-      );
+  if (tableIds) {
```
Comment (uladzimirdev) [inline frontend/src/metabase/selectors/metadata.ts:280]: we actually didn't need that `if/else`

```
if (tableIds) {
  return tableIds.map(table => metadata.table(table)).filter(isNotNull);
}

if (schema.database && schema.database.getTables().length > 0) {
  return schema.database
      .getTables()
      .filter(table => table.schema_name === schema.name);
}

return Object.values(metadata.tables).filter(
    table => table.schema && table.schema.id === schema.id,
  );
```

**Pull request merged**